src/Interface.jl

```diff
@@ -11,15 +12,19 @@ placenames(plc::AbstractPlaces) = error("function not defined for this type")
 nthings(thg::AbstractThings) = error("function not defined for this type")
 thingnames(thg::AbstractThings) = error("function not defined for this type")
 
-nzrows(a::AbstractMatrix) = find(sum(a .> 0, 2) .> 0)
-nzcols(a::AbstractMatrix) = find(sum(a .> 0, 1) .> 0)
-nnz(a::AbstractArray) = sum(a .> 0)
+nzrows(a::AbstractMatrix) = LinearIndices(Compat.sum(a .> 0, dims=2))[findall(Compat.sum(a .> 0, dims=2) .> 0)]
```

Comment:
:scream:

Reply:
Mwahahaha....

Reply:
I actually think that `nzrows` and `nzcols` are an R'ism, and instead this should be an iterator. Something like 
```julia
struct NZrows 
    a::AbstractMatrix
    b::Int
end
nzrows(a::AbstractMatrix) = NZrows(a, size(a, dims = 1))

Base.iterate(nzr::NZrows, i = 0) 
    while(true)
            i > nrz.b && return nothing    
            i+=1
            sum(nrz.a[i,:]) == 0 && return i
      end
end
```
?
    end

Reply:
That looks fine in principle but actually, do we need them? I can see that `occurring()` and `occupied()` are useful but can we just dump the `nzcols()` and `nzrows()` and move things up, or is there some reason we need both? I'm just thinking that if we have an assemblage with more dimensions (e.g. a grid or even something with a timeline as we were discussing) then the `nzcols/nzrows` won't necessarily be the right names anyway, so the abstraction isn't necessarily helpful, and we can just overload the iterator for assemblages with those grids, etc. Or do you use `nzrows()` and `nzcols()` elsewhere?

PS So more like:

```julia
struct Occurring{Asm <: AbstractAssemblage} 
    asm::Asm
    count::Int
end
occurring(asm::Asm) where Asm <: AbstractAssemblage = Occurring{Asm}(asm, nthings(asm))

Base.iterate(occ::Occurring, i = 0) 
   while(true)
        i == occ.count && return nothing
        row = view(occurrences(occ.asm), i += 1, :)    
        sum(row) > 0 && return (row, i)
   end
end
```

Reply:
And then maybe you just need to overload `thingview()` for different layouts (depending on what you need it for).

Reply:
Well yes and no - yes I want something like that, and no - I also think I got a little confused in my explanation. 
So, yes I want iterators like that, I think they should start with `each` and be like `eachspecies(choose = :occurring)`or something like that and be iterators that give meaningfull objects.

But that's not why I have `nzrows`. That is to help in the production of views, so you can limit views to only the sites where species occur etc. In that way, you can e.g. get the mean temperature over the range of a family by making a view into the family, dropping the sites where the family does not occur, then taking the mean of the temperature in those sites:  
```julia
fam = view(asm, species = asm[:traits, :family] == "test", dropsites = true)
mean(fam[:envir, :temperature])
```

Reply:
I'd say for now, let's just go with a simple version `nzrows(a) = findall(vec(Compat.sum(a, dims = 2)).>0)` and then merge.

Reply:
Hi @richardreeve I'm going to merge this today, then edit to get the simpler version of the code snippet above, if you don't protest.